lib/SQL/Abstract/mysql.pm

```diff
@@ -98,6 +140,95 @@ sub _order_by {
   return $sql, @bind;
 }
 
+sub _order_by_chunks {
+  my ($self, $arg) = @_;
+  if ($arg =~ /->/) {
+    return $self->_json_extract($arg);
+  }
+  else {
+    return $self->SUPER::_order_by_chunks($arg);
+  }
+}
+
+sub _select_fields {
+  my ($self, $fields) = @_;
+  return $fields unless ref $fields eq 'ARRAY';
+
+  if (grep /->/, @$fields) {
+    return join ',', map { $self->_json_extract($_, 1) } @$fields;
+  }
+  else {
+    return join ',', map { $self->_quote($_) } @$fields;
+  }
+}
+
+sub _update_set_values {
+  my ($self, $data) = @_;
+
+  for my $k (sort grep /->/, keys %$data) {
+    my ($label, $path) = $k =~ /(.+)->(.*)/;
+    my $origin = $self->_quote($label);
+    puke "you can\'t update $label and its values in the same query"
+      unless ($data->{$label}->[0] || 'json') =~ /^json/i;
+    if (defined $data->{$k}) {
+      if ($data->{$label}->[0]) {
+        puke "you can\'t update and remove values of $label in the same query"
+          unless $data->{$label}->[0] =~ /^json_set/i;
+      }
+      else {
+        $data->{$label}->[0] = $path ? $self->_sqlcase('json_set(') . "$origin)" : '?';
+      }
+      my $placeholder;
+      if (ref $data->{$k}) {
```

Comment:
Is this symmetric with what you have at https://github.com/jhthorsen/mojo-mysql/pull/51/files#diff-45bdaf6ecc6b539751f37122c23c1f17R53 ?

`if (ref $v eq 'HASH') { ...`

Reply:
It's not symmetric and `if (ref $v eq 'HASH') { ...` is wrong because it doesn't accept array references (the top level of a document can be an array, as you have mentioned in your first comment).
At this place we process the value of `field->some.key` and we can expect that everything that is a reference has to be JSON encoded.